File: javascript/src/test-opensearch.js
#!/usr/bin/env node

/**
 * OpenSearch Verbindungstest für die Schulungsumgebung
 * OS-unabhängig: Windows, Linux, macOS
 * 
 * Dieses Programm testet die Verbindung zu OpenSearch und zeigt
 * grundlegende Cluster-Informationen an.
 */

import { Client } from '@opensearch-project/opensearch';
import dotenv from 'dotenv';
import path from 'path';
import { fileURLToPath } from 'url';
import { existsSync } from 'fs';

// ES Module __dirname equivalent
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * Lädt die Umgebungsvariablen aus der .env Datei
 * Sucht in: ../../.env (übergeordnetes Verzeichnis)
 */
function loadEnvironment() {
    const envPath = path.join(__dirname, '..', '..', '.env');

    if (existsSync(envPath)) {
        dotenv.config({ path: envPath });
        return envPath;
    }

    // Keine .env gefunden, verwende Systemumgebungsvariablen
    console.log('ℹ️  Hinweis: Keine .env Datei gefunden, verwende Systemumgebungsvariablen');
    return null;
}

// Lade Umgebungsvariablen
const envPath = loadEnvironment();

/**
 * Erstellt einen OpenSearch Client mit SSL-Unterstützung
 */
function createOpenSearchClient() {
    const host = process.env.OPENSEARCH_HOST;
    const user = process.env.OPENSEARCH_USER;
    const password = process.env.OPENSEARCH_PASSWORD;
    const verifyCerts = false;

    const client = new Client({
        node: host,
        auth: {
            username: user,
            password: password
        },
        ssl: {
            rejectUnauthorized: verifyCerts
        }
    });

    return client;
}

/**
 * Testet die Verbindung zu OpenSearch und zeigt Cluster-Status
 */
async function testOpenSearchConnection() {
    console.log('='.repeat(60));
    console.log('OPENSEARCH VERBINDUNGSTEST');
    console.log(`Schulungsumgebung - JavaScript (${process.platform})`);
    console.log('='.repeat(60));
    
    try {
        // Erstelle OpenSearch Client
        const client = createOpenSearchClient();
        
        console.log('🔍 Teste OpenSearch Verbindung...');
        console.log(`Host: ${process.env.OPENSEARCH_HOST}`);
        console.log(`User: ${process.env.OPENSEARCH_USER}`);
        console.log('-'.repeat(50));
        
        // Teste Verbindung und hole Cluster-Info
        const info = await client.info();
        console.log('✅ Verbunden mit OpenSearch');
        console.log(`   Version: ${info.body.version.number}`);
        console.log(`   Cluster: ${info.body.cluster_name}`);
        console.log(`   Lucene Version: ${info.body.version.lucene_version}`);
        
        // Hole Cluster Health
        const health = await client.cluster.health();
        console.log(`🔍 Cluster-Zustand: ${health.body.status.toUpperCase()}`);
        console.log(`   Anzahl Nodes: ${health.body.number_of_nodes}`);
        console.log(`   Anzahl Daten-Nodes: ${health.body.number_of_data_nodes}`);
        
        // Teste Index-Liste
        const indices = await client.cat.indices({ format: 'json' });
        console.log(`📋 Anzahl Indices: ${indices.body.length}`);
        
        if (indices.body.length > 0) {
            console.log('   Verfügbare Indices:');
            indices.body.slice(0, 5).forEach(index => {
                console.log(`     - ${index.index} (${index['docs.count']} Dokumente)`);
            });
            if (indices.body.length > 5) {
                console.log(`     ... und ${indices.body.length - 5} weitere`);
            }
        }
        
        console.log('\n🎉 OpenSearch Verbindungstest erfolgreich!');
        process.exit(0);
        
    } catch (error) {
        console.error('❌ Fehler bei OpenSearch Verbindung:');
        console.error(`   ${error.message}`);
        console.error('\n💡 Mögliche Lösungen:');
        console.error('   1. Stellen Sie sicher, dass OpenSearch läuft (docker-compose -f docker-compose-os.yml up)');
        console.error('   2. Prüfen Sie die .env Datei im Projektverzeichnis');
        console.error('   3. Prüfen Sie die Netzwerkverbindung');
        console.error('   4. Warten Sie bis OpenSearch vollständig gestartet ist (kann 1-2 Minuten dauern)');
        
        process.exit(1);
    }
}

// Führe Test aus
testOpenSearchConnection();

File: python/src/test_opensearch.py
#!/usr/bin/env python3
"""
OpenSearch Verbindungstest für die Schulungsumgebung
OS-unabhängig: Windows, Linux, macOS
"""

import os
import sys
from pathlib import Path
from opensearchpy import OpenSearch
from dotenv import load_dotenv

def load_environment():
    """
    Lädt die Umgebungsvariablen aus der .env Datei
    Sucht in: ../../.env (übergeordnetes Verzeichnis)
    """
    script_dir = Path(__file__).parent
    env_path = script_dir.parent.parent / '.env';
    
    if env_path.exists():
        load_dotenv(dotenv_path=env_path)
        return str(env_path)
    
    # Keine .env gefunden, verwende Systemumgebungsvariablen
    print("ℹ️  Hinweis: Keine .env Datei gefunden, verwende Systemumgebungsvariablen")
    return None

# Lade Umgebungsvariablen
env_file = load_environment()

def test_opensearch_connection():
    """Testet die Verbindung zu OpenSearch und zeigt Cluster-Status"""
    
    try:
        # Erstelle OpenSearch Client
        client = OpenSearch(
            hosts=[os.environ["OPENSEARCH_HOST"]],
            http_auth=(os.environ["OPENSEARCH_USER"], os.environ["OPENSEARCH_PASSWORD"]),
            use_ssl=True,
            verify_certs=False,
            ssl_assert_hostname=False,
            ssl_show_warn=False
        )
        
        print("🔍 Teste OpenSearch Verbindung...")
        print(f"Host: {os.environ['OPENSEARCH_HOST']}")
        print(f"User: {os.environ['OPENSEARCH_USER']}")
        print("-" * 50)
        
        # Teste Verbindung und hole Cluster-Info
        info = client.info()
        print(f"✅ Verbunden mit OpenSearch")
        print(f"   Version: {info['version']['number']}")
        print(f"   Cluster: {info['cluster_name']}")
        print(f"   Lucene Version: {info['version']['lucene_version']}")
        
        # Hole Cluster Health
        health = client.cluster.health()
        print(f"🔍 Cluster-Zustand: {health['status'].upper()}")
        print(f"   Anzahl Nodes: {health['number_of_nodes']}")
        print(f"   Anzahl Daten-Nodes: {health['number_of_data_nodes']}")
        
        # Teste Index-Liste
        indices = client.cat.indices(format='json')
        print(f"📋 Anzahl Indices: {len(indices)}")
        
        if indices:
            print("   Verfügbare Indices:")
            for idx in indices[:5]:  # Zeige nur die ersten 5
                print(f"     - {idx['index']} ({idx['docs.count']} Dokumente)")
            if len(indices) > 5:
                print(f"     ... und {len(indices) - 5} weitere")
        
        print("\n🎉 OpenSearch Verbindungstest erfolgreich!")
        return True
        
    except Exception as e:
        print(f"❌ Fehler bei OpenSearch Verbindung:")
        print(f"   {str(e)}")
        print("\n💡 Mögliche Lösungen:")
        print("   1. Stellen Sie sicher, dass OpenSearch läuft (docker-compose -f docker-compose-os.yml up)")
        print("   2. Prüfen Sie die .env Datei im Projektverzeichnis")
        print("   3. Prüfen Sie die Netzwerkverbindung")
        print("   4. Warten Sie bis OpenSearch vollständig gestartet ist (kann 1-2 Minuten dauern)")
        return False

def main():
    """Hauptfunktion"""
    print("=" * 60)
    print("OPENSEARCH VERBINDUNGSTEST")
    print(f"Schulungsumgebung - Python ({sys.platform})")
    print("=" * 60)
    
    success = test_opensearch_connection()
    
    if success:
        sys.exit(0)
    else:
        sys.exit(1)

if __name__ == "__main__":
    main()
